### PR TITLE
Optimise picopass crypto to fix timing issues with newer readers.

### LINKED ISSUE #35: Picopass not properly emulating
### Describe the bug.

Picopass app emulation not working on Signo 40 (no reaction from the reader). I've been told that it is probably a timing issue with the auth.

@nvx helped diagnose

### Reproduction

1. read iclass card
2. save
3. emulate
4. no reaction from reader

### Target

_No response_

### Logs

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nvx**
# What's new

- Optimises the speed of the picopass crypto. This improves emulation compatibility.

# Verification 

- New versions should work on Signos and OmniKey readers that previously would not respond to the emulation.
- Resolves https://github.com/flipperdevices/flipperzero-good-faps/issues/35

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by hedger** [comment]
Neat! Got any actual comparisons of how massive the speedup is?

**Comment by nvx** [comment]
> Neat! Got any actual comparisons of how massive the speedup is?

I haven't benchmarked the crypto individually (the total frame delay time is all that maatters to a reader), but the frame delay time to respond to a `CHECK` command from the reader was previously 394 us and is now 288 us the same as most other commands (note that there is a delay in the code to prevent responses being sent in less than 288 us, so with that delay disabled it would be even faster, but responding too fast also causes issues).

Once it was as fast as the other commands I figured it was good enough so didn't bother looking any further go figure.

**Comment by skotopes** [comment]
@nvx @bettse how about version bump and release?)

**Comment by nvx** [comment]
> @nvx @bettse how about version bump and release?)

Sounds good to me. Want me to throw it into this PR? There's another tiny bug where when loading a saved card it defaults to the last item in the menu, should ideally slip a fix in for that too before pushing the new version (should be fixed by re-ordeding of the menu enum).

**Comment by skotopes** [comment]
Up to you guys, there is no fixed release schedule for this repository.

File: picopass/scenes/picopass_scene_saved_menu.c [hunk 1]
```diff
@@ -1,11 +1,11 @@
 #include "../picopass_i.h"
 
 enum SubmenuIndex {
-    SubmenuIndexDelete,
     SubmenuIndexInfo,
     SubmenuIndexWrite,
     SubmenuIndexEmulate,
     SubmenuIndexRename,
+    SubmenuIndexDelete,
```
Comment (bettse) [inline picopass/scenes/picopass_scene_saved_menu.c:8]: Thanks!

**Comment by nvx** [comment]
> Up to you guys, there is no fixed release schedule for this repository.

Done. Should be good to cut a new release now

**Comment by skotopes** [comment]
There is one more issue: if you open loclass from main menu you'll get HardFault

```(gdb) bt
#0  0x0800f79c in __furi_crash () at furi/core/check.c:166
#1  0x08006fa6 in MemManage_Handler () at firmware/targets/f7/furi_hal/furi_hal_interrupt.c:271
#2  <signal handler called>
#3  loclass_writer_write_start_stop (instance=0x0, start=start@entry=true) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/loclass_writer.c:47
#4  0x2001a3dc in picopass_worker_emulate (picopass_worker=0x20009540, loclass_mode=<optimized out>) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/picopass_worker.c:1239
#5  0x2001a520 in picopass_worker_task (context=0x20009540) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/picopass_worker.c:637
#6  0x08011c14 in furi_thread_body (context=0x2000a3e8) at furi/core/thread.c:90
#7  0x08011bd2 in furi_thread_catch () at furi/core/thread.c:61
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) f 3 
#3  loclass_writer_write_start_stop (instance=0x0, start=start@entry=true) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/loclass_writer.c:47
47	    bool write_success = stream_write_string(instance->file_stream, str);
(gdb) p instance 
$1 = (LoclassWriter *) 0x0
```

**Comment by nvx** [comment]
> There is one more issue: if you open loclass from main menu you'll get HardFault
> 
> ```
> #0  0x0800f79c in __furi_crash () at furi/core/check.c:166
> #1  0x08006fa6 in MemManage_Handler () at firmware/targets/f7/furi_hal/furi_hal_interrupt.c:271
> #2  <signal handler called>
> #3  loclass_writer_write_start_stop (instance=0x0, start=start@entry=true) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/loclass_writer.c:47
> #4  0x2001a3dc in picopass_worker_emulate (picopass_worker=0x20009540, loclass_mode=<optimized out>) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/picopass_worker.c:1239
> #5  0x2001a520 in picopass_worker_task (context=0x20009540) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/picopass_worker.c:637
> #6  0x08011c14 in furi_thread_body (context=0x2000a3e8) at furi/core/thread.c:90
> #7  0x08011bd2 in furi_thread_catch () at furi/core/thread.c:61
> Backtrace stopped: previous frame identical to this frame (corrupt stack?)
> (gdb) f 3 
> #3  loclass_writer_write_start_stop (instance=0x0, start=start@entry=true) at /Users/aku/Work/Flipper/flipperzero-good-faps/picopass/loclass_writer.c:47
> 47	    bool write_success = stream_write_string(instance->file_stream, str);
> (gdb) p instance 
> $1 = (LoclassWriter *) 0x0
> ```

I can't reproduce that (running the latest release firmware). It looks like it's failing to open the loclass log file on the SD card.

I threw in a commit to handle this case gracefully by showing an error on the screen rather than hard crashing, but it wouldn't resolve the actual underlying issue.

Do other functions that write to the SD card work? For example saving a read card.

**Comment by bettse** [comment]
Tested with both of my flippers and didn't get any crash when trying loclass.  Used my non-dev flipper and was able to successfully emulate against both my signo rp20 and signo rp40 👍 .

**Review by skotopes — APPROVED** [review]

**Comment by nvx** [comment]
sweet, do we need to do anything more to update the app catalog listing @skotopes? or is that taken care of because it's in the special good faps repo?

**Comment by skotopes** [comment]
@nvx Yes, just open PR with manifest update in catalog repo.

Also @bettse got permissions to do merge in this repo, so you don't have to wait for my pr approve.

**Comment by bettse** [comment]
😮 I saw getting added, but I hadn't been totally sure.  Exciting!

**Pull request merged**